File: python/src/arker/e2b.py
from __future__ import annotations

import shlex
from typing import Any

from ._compat import CommandResult, UnsupportedMixin, arker_create, arker_get, format_not_found, import_optional, run_command
from .computer import VM


class Commands:
    def __init__(self, vm: VM | None = None, native: Any | None = None) -> None:
        self._vm = vm
        self._native = native

    def run(self, command: str) -> CommandResult | Any:
        if self._vm is not None:
            return run_command(self._vm, command)
        return self._native.run(command)


class Files(UnsupportedMixin):
    _compat_name = "e2b"

    def __init__(self, vm: VM | None = None, native: Any | None = None) -> None:
        self._vm = vm
        self._native = native

    def read(self, path: str) -> str:
        if self._vm is None:
            return self._native.read(path)
        return self._vm.sync(path).decode("utf-8", "replace")  # type: ignore[union-attr]

    def write(self, path: str, content: str) -> None:
        if self._vm is None:
            result = self._native.write(path, content)
            return None if result is None else result
        self._vm.sync(path, content)
        return None

    def make_dir(self, path: str) -> None:
        if self._vm is None:
            if hasattr(self._native, "make_dir"):
                return self._native.make_dir(path)
            return self._native.makeDir(path)
        result = run_command(self._vm, f"mkdir -p {shlex.quote(path)}")
        if result.exit_code != 0:
            raise RuntimeError(result.stderr or f"mkdir failed with exit {result.exit_code}")
        return None

    def makeDir(self, path: str) -> None:
        return self.make_dir(path)

    def list(self, path: str) -> list[Any]:
        if self._vm is None:
            return self._native.list(path)
        quoted = shlex.quote(path)
        script = (
            f"cd {quoted} 2>/dev/null || exit 3; "
            "for e in * .*; do "
            '[ "$e" = "." ] && continue; [ "$e" = ".." ] && continue; [ -e "$e" ] || continue; '
            "if [ -d \"$e\" ]; then t=dir; else t=file; fi; "
            "printf '%s\t%s\n' \"$t\" \"$e\"; done"
        )
        result = run_command(self._vm, script)
        if result.exit_code != 0:
            raise RuntimeError(result.stderr or f"list failed with exit {result.exit_code}")
        return [
            {"type": line.split("\t", 1)[0], "name": line.split("\t", 1)[1]}
            for line in result.stdout.splitlines()
            if "\t" in line
        ]

    def exists(self, path: str) -> bool:
        if self._vm is None:
            return bool(self._native.exists(path))
        return run_command(self._vm, f"test -e {shlex.quote(path)}").exit_code == 0

    def remove(self, path: str) -> None:
        if self._vm is None:
            result = self._native.remove(path)
            return None if result is None else result
        result = run_command(self._vm, f"rm -rf {shlex.quote(path)}")
        if result.exit_code != 0:
            raise RuntimeError(result.stderr or f"remove failed with exit {result.exit_code}")
        return None


class Sandbox(UnsupportedMixin):
    _compat_name = "e2b"

    def __init__(self, vm: VM | None = None, native: Any | None = None) -> None:
        self._vm = vm
        self._native = native
        self.sandbox_id = vm.id if vm is not None else getattr(native, "sandbox_id", getattr(native, "id", None))
        self.sandboxId = self.sandbox_id
        self.commands = Commands(vm, getattr(native, "commands", None) if native is not None else None)
        self.files = Files(vm, getattr(native, "files", None) if native is not None else None)

    @staticmethod
    def _native_sandbox() -> Any | None:
        module = import_optional("e2b")
        return getattr(module, "Sandbox", None) if module is not None else None

    @classmethod
    def create(cls, template_id: str | None = None, **kwargs: Any) -> "Sandbox":
        timeout = kwargs.pop("timeout", kwargs.pop("timeout_ms", kwargs.pop("timeoutMs", None)))
        arker_config = kwargs.pop("arker", None)
        if kwargs:
            unsupported = ", ".join(sorted(kwargs))
            raise TypeError(f"Sandbox.create received unsupported option(s): {unsupported}")
        try:
            return cls(vm=arker_create(arker_config, template_id=template_id))
        except Exception:
            native = cls._native_sandbox()
            if native is None:
                raise
            if template_id is not None:
                return cls(native=native.create(template_id, timeout=timeout) if timeout is not None else native.create(template_id))
            return cls(native=native.create(timeout=timeout) if timeout is not None else native.create())

    @classmethod
    def connect(cls, sandbox_id: str, **kwargs: Any) -> "Sandbox":
        arker_config = kwargs.pop("arker", None)
        if kwargs:
            unsupported = ", ".join(sorted(kwargs))
            raise TypeError(f"Sandbox.connect received unsupported option(s): {unsupported}")
        vm = arker_get(sandbox_id, arker_config)
        if vm is not None:
            return cls(vm=vm)
        native = cls._native_sandbox()
        if native is not None and hasattr(native, "connect"):
            return cls(native=native.connect(sandbox_id))
        raise format_not_found("Sandbox", sandbox_id)

    def kill(self) -> None:
        if self._vm is not None:
            self._vm.delete()
            return
        if hasattr(self._native, "kill"):
            self._native.kill()
            return
        if hasattr(self._native, "close"):
            self._native.close()
            return
        raise AttributeError("Native E2B sandbox does not expose kill().")

File: typescript/src/e2b.ts
import { e2b as createE2BProvider } from "@computesdk/e2b";

import {
  assertSupportedObjectKeys,
  guardUnsupported,
  resolveTimeout,
  type CreateOptions,
  type UniversalSandbox,
} from "./compat/common.js";
import { createCompatSdk } from "./compat/runtime.js";

export interface CommandExitResult {
  stdout: string;
  stderr: string;
  exitCode: number;
}

const E2B_CREATE_OPTION_KEYS = ["timeout", "timeoutMs"] as const;

function sdk() {
  return createCompatSdk({
    fallbackProvider: createE2BProvider({}),
  });
}

export class Sandbox {
  readonly sandboxId: string;
  #box: UniversalSandbox;

  readonly commands: {
    run: (command: string) => Promise<CommandExitResult>;
  };

  readonly files: {
    read: (path: string) => Promise<string>;
    write: (path: string, content: string) => Promise<void>;
    makeDir: (path: string) => Promise<void>;
    list: (path: string) => Promise<unknown[]>;
    exists: (path: string) => Promise<boolean>;
    remove: (path: string) => Promise<void>;
  };

  private constructor(box: UniversalSandbox) {
    this.#box = box;
    this.sandboxId = box.sandboxId;
    this.commands = {
      run: async (command) => {
        const result = await this.#box.runCommand(command);
        return {
          stdout: result.stdout,
          stderr: result.stderr,
          exitCode: result.exitCode,
        };
      },
    };
    this.files = {
      read: (path) => this.#box.filesystem.readFile(path),
      write: async (path, content) => {
        await this.#box.filesystem.writeFile(path, content);
      },
      makeDir: async (path) => {
        await this.#box.filesystem.mkdir(path);
      },
      list: (path) => this.#box.filesystem.readdir(path),
      exists: (path) => this.#box.filesystem.exists(path),
      remove: async (path) => {
        await this.#box.filesystem.remove(path);
      },
    };
  }

  static async create(templateId?: string, opts?: CreateOptions): Promise<Sandbox> {
    if (templateId !== undefined && typeof templateId !== "string") {
      throw new Error(
        `Sandbox.create only supports create() or create(templateId, { timeoutMs }) in the Arker compatibility layer for e2b. ` +
          `Use the native e2b SDK for unsupported create signatures.`,
      );
    }
    assertSupportedObjectKeys(opts, E2B_CREATE_OPTION_KEYS, "e2b", "Sandbox.create");
    const timeout = resolveTimeout(opts);
    const box = await sdk().sandbox.create({
      ...(templateId ? { templateId } : {}),
      ...(timeout !== undefined ? { timeout } : {}),
    });
    return guardUnsupported(new Sandbox(box), "e2b");
  }

  static async connect(sandboxId: string): Promise<Sandbox> {
    const box = await sdk().sandbox.getById(sandboxId);
    if (!box) throw new Error(`Sandbox ${sandboxId} not found`);
    return guardUnsupported(new Sandbox(box), "e2b");
  }

  async kill(): Promise<void> {
    await this.#box.destroy();
  }
}
